implementations/rust/tcp_manager/src/tcp_worker.rs

```diff
@@ -0,0 +1,169 @@
+            Ok(mut tcp_len) => {
+                if tcp_len == 0 {
+                    return Ok(false);
+                }
+                let mut tcp_vec = tcp_buff[0..tcp_len].to_vec();
+                while tcp_vec.len() > 0 {
+                    // if self.message_length is 0, then decode the next byte(s) as message length
+                    if self.message_length == 0 {
+                        self.set_msg_len(&mut tcp_vec)?;
+                    }
+
+                    // we have a message length and an offset into the message buffer,
+                    // try to read enough bytes to fill the message
+                    let mut remaining_msg_bytes = self.message_length - self.offset;
+
+                    if tcp_vec.len() < remaining_msg_bytes {
+                        // not enough bytes to complete message, copy what there is and return
+                        self.message[self.offset..(self.offset + tcp_vec.len())]
+                            .clone_from_slice(&tcp_vec);
+                        self.offset += tcp_vec.len();
+                        return Ok(false);
+                    }
+
+                    // we have a complete message, route it
+                    let bytes_to_clone = self.message_length - self.offset;
+                    self.message[self.offset..self.message_length]
+                        .clone_from_slice(&tcp_vec[0..bytes_to_clone]);
+                    tcp_vec = tcp_vec.split_off(bytes_to_clone);
+                    self.decode_and_route_message(enqueue_message_ref.clone())?;
+                    self.offset = 0;
+                    self.message_length = 0;
+                }
+                Ok(true)
+            }
+            Err(e) => match e.kind() {
+                io::ErrorKind::WouldBlock => Ok(true),
+                _ => Err("***tcp receive failed".to_string()),
+            },
+        }
+    }
+}
+
+impl TcpWorker {
+    pub fn new_connection(stream: TcpStream) -> Self {
+        TcpWorker {
+            stream,
+            message: [0u8; MAX_MESSAGE_SIZE],
+            offset: 0,
+            message_length: 0,
+        }
+    }
+
+    fn set_msg_len(&mut self, varint: &mut Vec<u8>) -> Result<(), String> {
+        if let Ok((l, b)) = u16::decode(varint) {
+            self.message_length = l as usize;
+            varint.remove(0);
+            if varint_size(l) == 2 {
+                varint.remove(0);
+            }
+            Ok(())
+        } else {
+            Err("seg_msg_len failed".into())
+        }
+    }
+
+    fn decode_and_route_message(
+        &mut self,
+        enqueue_message_ref: Rc<RefCell<dyn EnqueueMessage>>,
+    ) -> Result<bool, String> {
+        match Message::decode(&self.message[0..self.message_length]) {
+            Ok((mut m_decoded, _)) => {
+                // fix up return tcp address with nat-ed address
+                let tcp_return = Address::TcpAddress(self.stream.peer_addr().unwrap());
```

Comment:
Is it assumed that `unwrap` will not fail? Any reason why to use `unwrap` vs `?`

Reply:
This can happen in a later PR too.